src/Request.ts

```diff
@@ -0,0 +1,350 @@
+		this.descendingPrefix = this.descendingPrefix || '-';
+		this.accepts = this.accepts || 'application/json';
+		this.fetch = this.fetch || function (kwArgs?: dstore.FetchArgs) {
+			const results = this._request(kwArgs);
+			return <dstore.FetchPromise<any>> QueryResults(results.data, {
+				response: results.response
+			});
+		};
+	}
+
+	protected _request(kwArgs: dstore.FetchArgs = {}): {
+		data: dstore.FetchPromise<T>;
+		total: Promise<number>;
+		response: ResponsePromise<T>
+	} {
+		// perform the actual query
+		const RequestHeader: { [ name: string ]: string } = null;
+		const headers = <typeof RequestHeader> lang.mixin(Object.create(this.headers), { Accept: this.accepts });
+
+		if ('headers' in kwArgs) {
+			lang.mixin(headers, kwArgs.headers);
+		}
+
+		const requestUrl = this._renderUrl(kwArgs.queryParams);
+
+		const response = request(requestUrl, {
+			method: 'GET',
+			headers: headers
+		});
+		const collection = this;
+		const parsedResponse = response.then(function (response: Response<string>) {
+			return collection.parse(response.data);
+		});
+		return {
+			data: parsedResponse.then(function (data) {
+				// support items in the results
+				const results = data.items || data;
+				for (let i = 0, l = results.length; i < l; i++) {
+					results[ i ] = collection._restore(results[ i ], true);
+				}
+				return results;
+			}),
+			total: parsedResponse.then(function (data) {
+				// check for a total property
+				const total = data.total;
+				if (total > -1) {
+					// if we have a valid positive number from the data,
+					// we can use that
+					return total;
+				}
+				// else use headers
+				return response.then(function (response) {
+					const rangeHeader = response.getHeader('Content-Range');
+					let rangeMatch: RegExpMatchArray;
+					let range: number;
+					if (rangeHeader && (rangeMatch = rangeHeader.match(/\/(.*)/)) && (range = Number(rangeMatch[ 1 ]))) {
+						return range;
+					} else {
+						return null;
+					}
+				});
+			}),
+			response: response
+		};
+	}
+
+	/**
+	 * Constructs filter-related params to be inserted into the query string
+	 *
+	 * @param filter The filter to render as part of a query string
+	 * @returns Filter-related params to be inserted in the query string
+	 */
+	protected _renderFilterParams(filter: Filter): string[] {
+		const type = filter.type;
+		const args = filter.args;
+		if (!type) {
+			return [ '' ];
+		}
+		if (type === 'string') {
+			return [ args[ 0 ] ];
+		}
+		if (type === 'and' || type === 'or') {
+			const joinToken = type === 'and' ? '&' : '|';
+			const renderedArgs = args.map(function (arg) {
+				// render each of the arguments to and or or, then combine by the right operator
+				const renderedArg = this._renderFilterParams(arg);
+				return ((arg.type === 'and' || arg.type === 'or') && arg.type !== type) ?
+					// need to observe precedence in the case of changing combination operators
+				'(' + renderedArg + ')' : renderedArg;
```

Comment:
This line should be indented one more level.

Otherwise, this section overall is definitely more readable than it was before. :+1: